### PR TITLE
Add SOAP task and transformer

### PR DESCRIPTION
**Pull request by mveyrenc**
Add a SOAP task and a SOAP transformer toi make SOAP request if PHP SOAP extension is loaded

### CODE REVIEW

File: DependencyInjection/CleverAgeProcessExtension.php [hunk 1]
```diff
@@ -34,6 +36,13 @@ public function load(array $configs, ContainerBuilder $container)
     {
         parent::load($configs, $container);
 
+        if (extension_loaded('soap')) {
+            $refl = new \ReflectionClass($this);
+            $loader = new ServiceLoader($container);
+            $serviceFolderPath = \dirname($refl->getFileName(), 2).'/Resources/config/services-soap';
```
Comment (VincentChalnot) [inline DependencyInjection/CleverAgeProcessExtension.php]: Why not using `__DIR__` ?

File: Soap/Client/Client.php [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+/**
+ * Class Client
+ *
+ * @author Madeline Veyrenc <mveyrenc@clever-age.com>
+ */
+class Client implements ClientInterface
+{
+    /**
+     * @var
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: `@var string`
Why not using inline comments ?
`/** @var string */`

File: Soap/Client/Client.php [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+     */
+    private $code;
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: Why not protected ? It would be nice to be able to extend this class if needed

File: Soap/Client/Client.php [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+     * @var string
+     */
+    private $lastResponseHeaders;
+
+    /**
+     * Client constructor.
+     *
+     * @param LoggerInterface $logger
+     */
+    public function __construct(LoggerInterface $logger)
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: If the logger is the only property that is always not null, all the other properties should be nullable

File: Soap/Client/Client.php [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * {@inheritdoc}
+     * @throws \UnexpectedValueException
+     */
+    public function getCode(): string
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: Why not putting code in the constructor if it can't be null ?

File: Soap/Client/Client.php [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+            throw new \UnexpectedValueException('Client code is not defined');
+        }
+
+        return $this->code;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setCode(?string $code): void
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: if code is mandatory, you can drop the `?`

File: Soap/Client/Client.php [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    {
+        $this->code = $code;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getWsdl(): string
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: If wsdl can be null, return type should be `?string`

Comment (VincentChalnot) [reply Soap/Client/Client.php]: Same for all other properties

File: Soap/Client/Client.php [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * @return \SoapClient|null
+     */
+    public function getSoapClient(): ?\SoapClient
+    {
+        return $this->soapClient;
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: Why not putting the soap client in the constructor ?

File: Soap/Client/Client.php [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+     */
+    public function getLogger(): LoggerInterface
+    {
+        return $this->logger;
+    }
+
+    /**
+     * @param LoggerInterface $logger
+     */
+    public function setLogger(LoggerInterface $logger): void
```
Comment (VincentChalnot) [inline Soap/Client/Client.php]: Why allowing to set the logger if it's already injected in the constructor?

File: Soap/Client/ClientInterface.php [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+*/
+
+namespace CleverAge\ProcessBundle\Soap\Client;
+
+/**
+ * Interface ClientInterface
+ *
+ * @author Madeline Veyrenc <mveyrenc@clever-age.com>
+ */
+interface ClientInterface
```
Comment (VincentChalnot) [inline Soap/Client/ClientInterface.php]: Report changes from the implementation to the interface if needed

File: Soap/Exception/MissingClientException.php [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ * Exception thrown when trying to fetch a missing Soap client
+ *
+ * @author Madeline Veyrenc <mveyrenc@clever-age.com>
+ */
+class MissingClientException extends \UnexpectedValueException implements ProcessExceptionInterface
+{
+    /**
+     * @param string $code
+     */
+    public function __construct($code)
```
Comment (VincentChalnot) [inline Soap/Exception/MissingClientException.php]: It's best not to change the default constructor of the base exception to allow the injection of the previous error if needed.
Setting a default message can be done with a static constructor method.

File: Soap/Client/Client.php (hunk 3 above)
Comment (mveyrenc) [reply Soap/Client/Client.php]: To force child classes to use getters and setters so it's easier to add deprecated on attribute

File: Soap/Client/Client.php (hunk 8 above)
Comment (mveyrenc) [reply Soap/Client/Client.php]: Some time you need to add some options before creating the client, authentification for exemple.
Most of the time, this class will be override to add authentification, token generation, etc.

File: Soap/Exception/MissingClientException.php (hunk 11 above)
Comment (mveyrenc) [reply Soap/Exception/MissingClientException.php]: So fix need to be apply on MissingTransformerException

Comment (mveyrenc) [reply Soap/Exception/MissingClientException.php]: and on other Exception of the process bundle

**Pull request merged**